src/Library/Fsm/Fsm.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <Utility/String/TransparentFunctors.h>
+#include <Library/Logger/LogCategory.h>
+
+#include "FsmEventHandler.h"
+#include "FsmState.h"
+
+#include <memory>
+#include <functional>
+#include <string>
+#include <vector>
+#include <unordered_map>
+
+struct FsmTransitionTarget {
+    std::string stateName;
+    std::function<bool()> condition{};
+};
+
+using FsmTransitions = std::unordered_map<TransparentString, std::vector<FsmTransitionTarget>, TransparentStringHash, TransparentStringEquals>;
+
+class Fsm : public FsmEventHandler {
+ public:
+    /**
+     * @brief Updates the current state of the Fsm or executes any pending transitions.
+     */
+    void update();
+
+    /*
+    * @brief Check if the Fsm has reached its internal Fsm::exitState.
+    */
+    [[nodiscard]] bool hasReachedExitState() const;
+
+    /**
+     * @brief Sets the next state to be reached in the Fsm. The transition does not occur immediately but will be executed during the next Fsm::update() call.
+     * The jumpToState function allows unconditional transitions without requiring a previously defined transition connecting the current state to the target state.
+     * Subsequent calls to Fsm::jumpToState or Fsm::scheduleTransition will overwrite the target state since the actual transition happens during the next Fsm::update() call.
+     * @param stateName The name of the state to transition to. This name must belong to a state that has been previously added through Fsm::addState.
+     */
+    void jumpToState(std::string_view stateName);
+
+    /**
+     * Global string used to specify the state used to exit from the Fsm.
+     * When the Fsm jumps to this state, Fsm::hasReachedExitState() will return true.
+     */
+    static const std::string_view exitState;
+
+    struct StateEntry {
+        std::string name;
+        std::unique_ptr<FsmState> state;
+        FsmTransitions transitions;
+    };
+
+    void addState(std::unique_ptr<StateEntry> stateEntry);
+
+    static const LogCategory fsmLogCategory;
+
+ private:
+    // FsmEventHandler implementation
+    virtual bool event(const PlatformEvent *event) override;
+
+    void _performPendingTransition();
+    void _updateCurrentState();
+    void _performAction(FsmAction &action);
+    void _scheduleTransition(std::string_view transition);
+    StateEntry *_getStateByName(std::string_view stateName);
+
+    std::unordered_map<TransparentString, std::unique_ptr<StateEntry>, TransparentStringHash, TransparentStringEquals> _states;
+    StateEntry *_currentState{};
+    StateEntry *_nextState{};
```

Comment:
Methods in this class are communicating with each other through a shared `_nextState` member variable. I haven't tried to rewrite the code here myself, but my bet would be that we can get rid of this field, and the logic will become simpler.

The reason for having it as a field is b/c we expose `jumpToState` in the public API, which I think we shouldn't do.  Just set the start state in builder (that's the most logical place for it anyway), drop `jumpToState` from the public API (so that Fsm actually would have proper guarantees – jumping between states requres and edge), and then I think other things will fall into place nicely.

I'm OK if you do this in a separate PR (this review has been going on for some time already), but at least leave a TODO.

Reply:
I'm going to write down my out loud thoughts just in case something doesn't feel right:
`jumpToState` is also used in future branches to re-enter inside the fsm. But those branches are not based on the new updates I made based on your feedbacks.
To give you a bit more context:
In other branches I have the video sequence and all the main menu states ( party creation/main menu/credits ) in one `Fsm`. If we need to go back to main menu from the game ( by pressing exit from the pause menu ) or jumping to the party creation ( by pressing the new game from the pause menu ) we need to go back to the `Fsm` and jump to the right state.
But now I could just solve the problem differently. The `StartState` I already created in this branch can take care of that and trigger the right transition instead. By doing so we can always have one `entry point/start state` and the `jumpToState` is not required anymore.

I like the idea of not postponing the transition to the next update call too. 
Thanks to the `FSMAction` we can now be sure that a transition has been requested when the `enter/update` methods are done. With the original implementation I had a `scheduleTransition` which could be call at any time, creating a potential discrepancy between the `FSMState::exit()` call and whatever is still being executed after the `scheduleTransition` call.

In few words, I'm going to do what you're asking :D

Reply:
Yeah I think the system you have in place is already expressive enough, given that you can add conditions on edges,  AND can pick the edge from inside the enter() / update().

Reply:
jumpToState and nextState gone. Now the transitions are executed immediately.